bitcoin/src/network/constants.rs

```diff
@@ -49,19 +52,20 @@ use crate::internal_macros::user_enum;
 /// 60001 - Support `pong` message and nonce in `ping` message
 pub const PROTOCOL_VERSION: u32 = 70001;
 
-user_enum! {
-    /// The cryptocurrency to act on
-    #[derive(Copy, PartialEq, Eq, PartialOrd, Ord, Clone, Hash, Debug)]
-    pub enum Network {
-        /// Classic Bitcoin
-        Bitcoin <-> "bitcoin",
-        /// Bitcoin's testnet
-        Testnet <-> "testnet",
-        /// Bitcoin's signet
-        Signet <-> "signet",
-        /// Bitcoin's regtest
-        Regtest <-> "regtest"
-    }
+/// The cryptocurrency network to act on.
+#[derive(Copy, PartialEq, Eq, PartialOrd, Ord, Clone, Hash, Debug)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
+#[cfg_attr(feature = "serde", serde(rename_all = "lowercase"))]
+pub enum Network {
```

Comment:
`#[non_exhaustive]`

Reply:
Sure thing.

Reply:
:shrug: "Remove user_enum macro" is the MR and commit message..

Reply:
Heads up that the "complaining about `#[non_exhaustive]` on `Network`" issue is here https://github.com/rust-bitcoin/rust-bitcoin/issues/2225 and we basically agreed to eliminate any direct use of the enum in the library, and then move the enum itself to another crate where it won't cause breakage (primarily because nobody will use it..) and then we can remove `#[non_exhaustive]`.

Instead this library will receive network params by taking generic `Into<Params>` or `AsRef<Params>` arguments, and you can make your own network enum that implements this and describes exactly the set of networks that you support.

Reply:
To be fair, after all the drama introducing this change has caused I was surprised to see it come in so nonchalantly. I will admit that this should have been in a stand alone PR, the repercussions discussed, and a mention in the changelog. I take responsibility for the lack of all three.

Reply:
I think this is one of those cases where nobody could predict how much impact it would have. It seemed totally obvious that the enum should be `non_exhaustive`. Well, it kinda still is obvious to me but I see how *also* having other shared enums helps people not write a few lines of code. Though now that I think about it, it's super ironic that people here go as far as to NIH something just to avoid dependency  and they refuse to write an enum...

To be honest, there were multiple occasions where I was tempted to say "if you don't want strict API, don't use Rust". But I'm pretty sure it'd come off wrong even though I honestly believe Rust is not that useful if you want to write a quick dirty thing and great if you write something security-critical.

Reply:
Rust's lack of dependent types makes it basically impossible to write many of the APIs that we want (in particular APIs surrounding `Network`, e.g. addresses whose network is tied to them but which can still be parsed as the same type). Its coherence rules are constantly giving us trouble when we try to break crates apart. Rust's lack of inheritance or monkeypatching or anything *also* makes it impossible to NIH certain things, especially traits but sometimes enums whose job it is to provide interop. So "just write an enum" is not a solution to a lack of a usable `Network` enum. (But IMHO it *is* a solution if library crates avoid referring directly to the enum and instead use their interop with `Params`, since this does enable people to write their own enums.)

So if I wanted a strict API I'd use Idris or something, but Rust is definitely a language for getting work done.

Philosophy aside -- I was also caught off-guard by what a pain in the ass `#[non_exhaustive]` turned out to be, since it seemed unlikely that people were supporting the existing full set of networks and that they were doing so by exhaustively matching. I think my own code pretty consistently just panics on some networks.

Reply:
> Rust's lack of dependent types makes it basically impossible to write many of the APIs that we want

True, unfortunately.

> Its coherence rules are constantly giving us trouble when we try to break crates apart.

Sadly, this may be unsolvable. The rules protect against other problems. I was thinking about it many times but couldn't come up with a reasonable solution.

> Rust's lack of inheritance

Thank God for it! IME inheritance is just a mess. Though it's true that Rust would benefit from a delegation syntax. There were multiple attempts at RFC but sadly it doesn't seem like it's going anywhere anytime soon.

> So if I wanted a strict API I'd use Idris or something

I'd love to but last time I checked Idris was experimental and IDK how good performance of it is.